File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-3.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size() - 1;
        int mid = r / 2;

        while (true) {
            cout << "Left: " << nums[l] << ", Mid: " << nums[mid] << ", Right: " << nums[r] << "\n";
            if (nums[mid] == target) return mid;
            if (nums[l] == target) return l;
            if (nums[r] == target) return r;
            if (r - l < 2) break;

            if (nums[l] < nums[mid]) {
                if (nums[l] <= target && nums[mid] >= target) {
                    r = mid;
                } else {
                    l = mid + 1;
                }
            }
            else if (nums[r] > nums[mid]) {
                if (nums[mid] <= target && nums[r] >= target) {
                    l = mid;
                } else {
                    r = mid - 1;
                }
            }

            mid = (l + r) / 2;
        }

        return -1;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-3.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        comp = {}

        for i in range(len(nums)):
            cur_comp = 0 - nums[i]
            if cur_comp in comp:
                comp[cur_comp].append(i)
            else:
                comp[cur_comp] = [i]

        left = 0
        right = len(nums)
        res = []
        used = set()

        while left < len(nums) - 2:
            while right > left + 1:
                right -= 1

                cur_sum = nums[left] + nums[right]

                if cur_sum in comp:
                    
                    third_idx = next((i for i in comp[cur_sum] if i != left and i != right), None)
                    if third_idx is None:
                        continue
                    
                    triplet = [nums[third_idx], nums[left], nums[right]]
                    triplet_set = frozenset(triplet)

                    if triplet_set in used:
                        continue

                    res.append(triplet)
                    used.add(triplet_set)

            left += 1
            right = len(nums)
        
        return res